src/emulator/modules/SceLibKernel/SceLibKernel.cpp

```diff
@@ -32,8 +34,164 @@
+    if (!mutex) {
+        return error("sceKernelLockMutex", SCE_KERNEL_ERROR_UNKNOWN_MUTEX_ID);
+    }
+
+    if (LOG_SYNC_OBJECTS) {
+        LOG_DEBUG("Locking mutex: uid:{} thread_id:{} name:\"{}\" attr:{} lock_count:{}",
+            mutexid, thread_id, mutex->name, mutex->attr, mutex->lock_count);
+    }
+
+    const std::unique_lock<std::mutex> lock(mutex->mutex);
+
+    const ThreadStatePtr thread = lock_and_find(thread_id, host.kernel.threads, host.kernel.mutex);
+
+    bool is_recursive = (mutex->attr & SCE_KERNEL_MUTEX_ATTR_RECURSIVE);
+
+    // if (is_lw) {
+
+    if (mutex->lock_count > 0) {
+        // Already owned
+
+        if (mutex->owner == thread) {
+            // Owned by ourselves
+
+            if (is_recursive) {
+                // TODO: Recursive mutex support
+                LOG_ERROR("Recursive mutexes unsupported.");
+                return -1;
+            } else {
+                return SCE_KERNEL_ERROR_LW_MUTEX_RECURSIVE;
+            }
+
+        } else {
+            // Owned by someone else
+            // Sleep thread!
+
+            const std::unique_lock<std::mutex> lock2(thread->mutex);
+            assert(thread->to_do == ThreadToDo::run);
+            thread->to_do = ThreadToDo::wait;
+            mutex->waiting_threads.push_back(thread);
+
+            stop(*thread->cpu);
+        }
+    } else {
+        // Not owned
+        // Take ownership!
+
+        mutex->lock_count += lock_count;
+        mutex->owner = thread;            
+    }
+
+    return SCE_KERNEL_OK;
+}
+
+int unlock_mutex(HostState& host, SceUID thread_id, MutexPtrs& host_mutexes, SceUID mutexid, int unlock_count) {
+    assert(unlock_count == 1);
+
+    const MutexPtr mutex = lock_and_find(mutexid, host_mutexes, host.kernel.mutex);
+    if (!mutex) {
+        return error("sceKernel(Lw)UnlockMutex", SCE_KERNEL_ERROR_UNKNOWN_MUTEX_ID);
+    }
+
+    if (LOG_SYNC_OBJECTS) {
+        LOG_DEBUG("Unlocking mutex: uid:{} thread_id:{} name:\"{}\" attr:{} lock_count:{}",
+            mutexid, thread_id, mutex->name, mutex->attr, mutex->lock_count, unlock_count);
+    }
+
+    const std::unique_lock<std::mutex> lock(mutex->mutex);
+
+    const ThreadStatePtr cur_thread = lock_and_find(thread_id, host.kernel.threads, host.kernel.mutex);
+    if (cur_thread == mutex->owner) {
+        if (unlock_count > mutex->lock_count) {
+            return error("sceKernel(Lw)UnlockMutex", SCE_KERNEL_ERROR_LW_MUTEX_UNLOCK_UDF);
+        }
+        mutex->lock_count -= unlock_count;
+        if (mutex->lock_count == 0) {
+            mutex->owner = nullptr;
+            if (mutex->waiting_threads.size() > 0) {
+                const ThreadStatePtr thread = mutex->waiting_threads.back();
+                assert(thread->to_do == ThreadToDo::wait);
+                thread->to_do = ThreadToDo::run;
+                mutex->waiting_threads.pop_back();
+                mutex->lock_count++;
```

Comment:
this should be saved lock_count in queue

Reply:
indeed